Add CodeMirror to Additional CSS / Custom HTML block

okmttdhr opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?
<!-- In a few words, what is the PR actually doing? -->
This PR introduces inline code completion and linting into Global Style's Additional CSS and Custom HTML block by using CodeMirror.

### Additional CSS

before

https://github.com/WordPress/gutenberg/assets/5287479/ca7a14dc-6ddf-4b2e-b097-bfb7b3abb70a

after

https://github.com/WordPress/gutenberg/assets/5287479/2f0ff26b-ea59-4432-80b1-181bd2cfa520

<details><summary>previous commits</summary>
<p>

https://github.com/WordPress/gutenberg/assets/5287479/b48b2bba-d5be-4fb0-a685-993feb8e823e

</p>
</details> 

### Custom HTML

before

https://github.com/WordPress/gutenberg/assets/5287479/d94011c9-99f6-4ad4-a833-73439550a16f

after

https://github.com/WordPress/gutenberg/assets/5287479/1e859e9b-3489-410b-bd66-295bba9cb7d4

### Chunks are loaded on demand and reused

https://github.com/WordPress/gutenberg/assets/5287479/960ceef3-caaf-4d8c-9a5c-e8f387ba5ceb

## Why?
<!-- Why is this PR necessary? What problem is it solving? Reference any existing previous issue(s) or PR(s), but please add a short summary here, too -->

CodeMirror is first introduced to Gutenberg in this PR: https://github.com/WordPress/gutenberg/pull/4348 but removed in this PR: https://github.com/WordPress/gutenberg/pull/10396. The follow-up issue: https://github.com/WordPress/gutenberg/issues/10423 was created, but there haven't been any further updates. 

This PR addresses https://github.com/WordPress/gutenberg/issues/10423 and https://github.com/WordPress/gutenberg/issues/47945.

## How?
<!-- How is your PR addressing the issue at hand? What are the implementation details? -->
Branched from PR https://github.com/WordPress/gutenberg/pull/53380, this update utilizes Webpack’s dynamic `import()` as a temporary measure to lazy-load CodeMirror, as I agree with [this comment](https://github.com/WordPress/gutenberg/pull/53380#issuecomment-1759731167).

This code should work even when we adopt importmap later. Once we turn all `@wordpress/*` packages (specifically `block-editor` in this case) into modules by using `output.module=true`, we can externalize `codemirror` by using [`defaultRequestToExternalModule`](https://github.com/WordPress/gutenberg/blob/7b38263aea571c0c804c85b19ced92c6dfe6289b/packages/dependency-extraction-webpack-plugin/lib/util.js#L78) in `DependencyExtractionWebpackPlugin` and register the script as a module by [`wp_register_script_module('@wordpress/block-editor')`](https://github.com/WordPress/wordpress-develop/pull/5818) setting `codemirror` as a dependency module. 

## Testing Instructions

### Additional CSS

- Select the global styles panel on the right
- Make sure the dev tool network tab is open
- In the overflow menu top right, select Additional CSS
- Make sure the chunks get loaded on demand
- Add some CSS to the editor box and make sure line numbers appear, CSS autocompletion appears
- Make sure custom CSS is applied

### Custom HTML

- Make sure the dev tool network tab is open
- Add a Custom HTML block
- Make sure the chunks get loaded on demand
- Add some HTML to the editor and make sure line numbers appear, HTML autocompletion appears

okmttdhr committed:
Add codemirror to Additional CSS

okmttdhr committed:
Add codemirror to Custom HTML block

okmttdhr committed:
Fix styles

okmttdhr committed:
Add focus change effect

talldan commented on packages/block-library/package.json:
```diff
@@ -65,6 +66,7 @@
 		"@wordpress/wordcount": "file:../wordcount",
 		"change-case": "^4.1.2",
 		"classnames": "^2.3.1",
+		"codemirror": "6.0.1",
```
I wonder if some linting to prevent this being imported the traditional way would be beneficial to ensure codemirror isn't accidentally bundled in the future.

I think there are some restricted import statements already for particular packages, but I'm not familiar with how that's implemented. 🤔

okmttdhr replied:
Thanks for your suggestion :) I addressed it in https://github.com/WordPress/gutenberg/pull/60155/commits/421f2d4ba1256ad97198a2b8004d14af5915e970

jasmussen commented:
Thanks for the contribution. It may be good to search the archive for "codemirror" as well, there have been past conversations exploring adding it to the code editor section, which ultimately failed. It would be good to see what lessons could potentially be learned adding it here. I'd like to see this land, but there are some checks and balances that need to be considered. Perhaps an option even is to turn it on or off. This could be in preferences and/or onboarding, that might even let it be available for the code editor too.

A few notes from a superficial review, here's a GIF showing some light CSS editing and tabbing:

![tabbing into code mirror](https://github.com/WordPress/gutenberg/assets/1204802/97503f41-38f1-43be-800c-970e31e4b225)

* The input field needs a border in its resting state. Essentially it should look the same as before, a $gray-600 border with its corners set to $radius-block-ui
* The focus style should also be the same as before, a 1.5px thick spot color blue.
* Can the input field be full-height, same as before?
* Do we actually need to show the line-numbers? In this particular narrow context, they take up quite a bit of horizontal space. If not, we'd probably want to do some light visul changes to reduce the footprint.
* How hard would it be to change the syntax highlighting colors? It would be good to get closer to the core UI colors we have available, make it feel more at home and ensure good contrast.
* Can we use the same code font-stack for the field? It's $editor-html-font.

Probably related to the past conversations on codemirror, IMO the biggest potential benefit of using codemirror is tabbing to indent/outdent code. Yet that's also one of the main challenges, as we still need a way for keyboard users to enter and exit this field. That's something to consider the best practices for.

okmttdhr committed:
Add codemirror as restrictedImport

okmttdhr committed:
Ensure minimum lines to be full-width

okmttdhr committed:
Add indenting with tab

okmttdhr committed:
Add instructions to focus out of the editor

okmttdhr committed:
Align styles with the existing ones

okmttdhr commented:
Thank you for your review, @jasmussen!

>The input field needs a border in its resting state. Essentially it should look the same as before, a $gray-600 border with its corners set to $radius-block-ui
The focus style should also be the same as before, a 1.5px thick spot color blue.
Can the input field be full-height, same as before?
Can we use the same code font-stack for the field? It's $editor-html-font.

I have addressed these points accordingly 🙂 

>Do we actually need to show the line-numbers? In this particular narrow context, they take up quite a bit of horizontal space. If not, we'd probably want to do some light visul changes to reduce the footprint.

I think line numbers serve as a clear & common indicator for developers that "Oh, this is the place I can write code". However, I agree that the space is limited in terms of the Additional CSS input, so I've removed it, while retaining it for a Custom HTML block. 

>the biggest potential benefit of using codemirror is tabbing to indent/outdent code. Yet that's also one of the main challenges, as we still need a way for keyboard users to enter and exit this field. That's something to consider the best practices for.

I implemented tab-based indentation following the documentation: [CodeMirror Language Config Example](https://codemirror.net/examples/tab/). It mentions that using Tab for indentation is not enabled by default because it could conflict with [Success Criterion 2.1.2 No Keyboard Trap](https://www.w3.org/TR/WCAG21/#no-keyboard-trap). I attempted to address this by implementing Esc & Tab to focus out (which is the same behavior as focusing on/out blocks in the editor) and by adding instructions using `aria-describedby` (https://github.com/WordPress/gutenberg/pull/60155/commits/9394fc9b305ae18aa730a6229fe5e54e19585672) what are your thoughts on this? 

>How hard would it be to change the syntax highlighting colors? It would be good to get closer to the core UI colors we have available, make it feel more at home and ensure good contrast.

Technically, altering the syntax highlighting colors is straightforward — it's about overriding the original CSS. However, it's tricky to select the appropriate colors, IMO. For instance, the following CSS already contains some colors, necessitating consideration in determining which colors should be defined and used in which state (like when highlighting the text).

<img width="262" alt="Screenshot 2024-03-27 at 15 55 15" src="https://github.com/WordPress/gutenberg/assets/5287479/6ab0482f-ba1b-42a0-a3d6-a111c5f0017d">
<img width="262" alt="when highlighting 'o'" src="https://github.com/WordPress/gutenberg/assets/5287479/83415fdc-3552-496f-904a-aff8ef99c292">

>It may be good to search the archive for "codemirror" as well, there have been past conversations exploring adding it to the code editor section, which ultimately failed. It would be good to see what lessons could potentially be learned adding it here.

As far as I know, CodeMirror was first introduced to Gutenberg in this PR: https://github.com/WordPress/gutenberg/pull/4348 but removed later in this PR: https://github.com/WordPress/gutenberg/pull/10396. It was mainly due to the difficulty of lazy loading at that time. Now, importmap is coming soon in WP 6.5, and it's ready to get prepared by using `import()`.

I'm aware there have been several UX discussions, but I may need more time to find/review them thoroughly. 🙂 Please let me know if anything can be improved.

- https://github.com/WordPress/gutenberg/pull/4803
- https://github.com/WordPress/gutenberg/pull/3195#issuecomment-340391042

okmttdhr committed:
Fix lint

okmttdhr commented:
> Perhaps an option even is to turn it on or off. This could be in preferences and/or onboarding, and it might even be made available for the code editor too.

I'm trying to understand your concerns more clearly. 
Are you worried about not providing users with the option to disable syntax highlighting, or are you concerned about shipping a feature that might not be perfect? If it's the former, are you suggesting that there could be a demand for an option to turn off syntax highlighting? If it's about the latter, considering an experimental flag to introduce this feature could be a viable approach. I'm interested in the reasoning behind your comment :)

jsnajdr commented on packages/block-editor/src/components/global-styles/advanced-panel.js:
```diff
@@ -37,40 +38,130 @@ export default function AdvancedPanel( {
+			const { EditorView, basicSetup } = await import( 'codemirror' );
+			const { indentWithTab } = await import( '@codemirror/commands' );
+			const { keymap } = await import( '@codemirror/view' );
+			const { css } = await import( '@codemirror/lang-css' );
```
Importing these individual packages will create four webpack chunks, and the loading will take more time than necessary, because the next load starts only after the previous one has finished.

Better to move the `EditorView` code into a separate module and load it all at once:
```js
const editorRef = useRef();
useEffect( () => {
  import( './code-mirror-view' ).then( ( { default: createEditor } ) => {
    createEditor( {
      parent: editorRef.current,
      mode: 'css', // or 'html'
      onChange: handleOnChange,
      onBlur: handleOnBlur,
    } );
  } );
}, [] );
```

Then there will be just one `code-mirror-view` chunk that's loaded on demand. There's a bit of trouble with the `lang-css` and `lang-html` modules. Do we want to bundle them both in a single chunk? That depends on how big they are.

Or the dynamic module can contain the React component that shows the editor: the `<div>` together with the `useRef` and the `useEffect` (this time the effect doesn't do dynamic `import`). Then it can be used like:
```js
const CodeEditor = React.lazy( () => import( './code-mirror-view' ) );

<Suspense fallback={ <Placeholder /> }>
  <CodeEditor id className aria-describedBy value onChange onBlur />
</Suspense>
```

It would be also a good idea to not load the CodeMirror editor not when the `AdvancedPanel` is rendered, that will load it too often. Load it only after clicking on the CSS field, only after the user shows a clear intent to edit.

okmttdhr replied:
Thank you for your review/insights, @jsnajdr! 😄 

>and the loading will take more time than necessary, because the next load starts only after the previous one has finished

Not exactly, in terms of **`import`**. Fundamentally, based on my understanding of [how modules are waited for and executed](https://github.com/tc39/proposal-top-level-await#semantics-as-desugaring), the following code

```php
import { a } from './a.mjs';
import { b } from './b.mjs';
import { c } from './c.mjs';

console.log(a, b, c);
```

would be roughly equivalent to the following: 

```php
import { promise as aPromise, a } from './a.mjs';
import { promise as bPromise, b } from './b.mjs';
import { promise as cPromise, c } from './c.mjs';

export const promise = Promise.all([aPromise, bPromise, cPromise]).then(() => {
	console.log(a, b, c);
});
```

At least, Webpack behaves in this manner, even for **`import()`**. Here’s a screenshot showing that the requests for chunks run **in parallel** when using the original approach.

<img width="1295" alt="Screenshot 2024-03-28 at 13 47 15" src="https://github.com/WordPress/gutenberg/assets/5287479/8763f12e-bf4e-467a-ad52-82d8cd95b8b4">

**However**, this does not seem to be stated explicitly in [the](https://github.com/tc39/proposal-dynamic-import) [spec](https://tc39.es/ecma262/multipage/ecmascript-language-expressions.html#sec-import-calls) of `import()`, and my tests confirmed that native dynamic imports execute requests **sequentially** both at the top level and within function scopes. I’ll address this later in this comment. 

>Importing these individual packages will create four webpack chunks

This also applies even if we:

- consolidate everything into one React component,
- AND use `React.lazy()`
- AND use static `import {} from 'codemirror'`

because we are [splitting the vendor's chunk](https://github.com/WordPress/gutenberg/blob/59a81b5666a1ee6bb69dbe115c177c1c3bc0d0b9/packages/scripts/config/webpack.config.js#L138). 

This approach does not create a single chunk, even with all logic extracted into a component and using static import (although we can tweak chunks further by using the [optimization.splitChunks](https://webpack.js.org/plugins/split-chunks-plugin/#optimizationsplitchunks) option).

Thus, the requests for the “one-component approach” looked like this, showing no difference from the original approach:

<img width="1294" alt="Screenshot 2024-03-28 at 13 55 10" src="https://github.com/WordPress/gutenberg/assets/5287479/eae4a729-ad14-4206-b393-e4d3220071f4">

**Given the above, I propose the following;**

- For explicit parallelization, use **`Promise.all([import(), import(), …])`** rather than fetching a single larger chunk at once. This ensures that requests run in parallel, even after transitioning to native dynamic import.
- Extract “EditorView”: I favor the one-component approach for its cohesiveness (rather than for performance reasons), so I encapsulated the logic into a component for reuse. 🙂

For now, parallel chunk requests seem to be more efficient.

>Then there will be just one `code-mirror-view` chunk that's loaded on demand. There's a bit of trouble with the `lang-css` and `lang-html` modules. Do we want to bundle them both in a single chunk? That depends on how big they are.

I don’t think it’s a good idea to load all **`lang-*`** modules in this component because we may want to support additional language modes in the future. (Even with the “parallel requests” strategy, this could potentially max out the browser's limit on simultaneous requests.)

>It would be also a good idea to not load the CodeMirror editor not when the AdvancedPanel is rendered, that will load it too often. Load it only after clicking on the CSS field, only after the user shows a clear intent to edit.

On a slow 3G network, this delay could be around 8 seconds. Making users wait that long for syntax highlighting isn't ideal, IMO.

I'd consider more eager loading strategies (if that doesn't interfere with user interaction). For instance, for Additional CSS, we could load the chunks when the three-dots menu is opened, or for a Custom HTML block, load them when the block appears in the search results. Alternatively, **`prefetch`** might be a viable approach. WDYT?

Once the chunks are loaded, subsequent network requests are unnecessary, even if users access Additional CSS multiple times. Thus, “loading them too often” shouldn’t pose an issue, in my opinion. We just need to be careful about how it affects users’ actual interactions/experiences.

jsnajdr replied:
> we are splitting the vendor's chunk

We should consider _not_ doing this. After all, currently Gutenberg doesn't do it on purpose, but only because it's a webpack default and we didn't do any dynamic chunks until now.

In my block lazy loading experiment I disabled it, see this comment: https://github.com/WordPress/gutenberg/pull/55585/files#r1383350711

It helps to keep a nice file structure with human-comprehensible names.

okmttdhr committed:
Extract EditorView component

okmttdhr committed:
Fix incorrect use of var()

okmttdhr committed:
Reuse EditorView component

mirka commented on packages/block-editor/package.json:
```diff
@@ -32,6 +32,10 @@
 		"@babel/runtime": "^7.16.0",
+		"@codemirror/commands": "6.3.3",
+		"@codemirror/lang-css": "6.2.1",
+		"@codemirror/lang-html": "6.2.1",
```
```suggestion
		"@codemirror/lang-html": "6.4.8",
```

jsnajdr replied:
Yes, this is a typo that leads to `npm install` error:
```
No matching version found for @codemirror/lang-html@6.2.1.
```

mirka commented on packages/edit-site/src/components/global-styles/style.scss:
```diff
@@ -119,24 +119,33 @@
+				box-shadow: 0 0 0 calc(1.5px - 1px) var(--wp-admin-theme-color);
+				border-color: var(--wp-admin-theme-color);
+				outline: transparent solid 2px;
+			}
```
Can we reuse these mixins here?

https://github.com/WordPress/gutenberg/blob/4cfaab6302c862ce264247ebef31b45729f04343/packages/base-styles/_mixins.scss#L74-L89

mirka commented on packages/block-editor/src/components/global-styles/advanced-panel.js:
```diff
@@ -37,40 +90,61 @@ export default function AdvancedPanel( {
 		);
 	}
 
+	const [ customCSS, setCustomeCSS ] = useState( '' );
```
```suggestion
	const [ customCSS, setCustomCSS ] = useState( '' );
```

mirka replied:
I think there's a bug where the Additional CSS editor will not show without already having an `inheritedValue`.

mirka commented on packages/block-library/src/html/edit.js:
```diff
@@ -73,11 +73,15 @@ export default function HTMLEdit( { attributes, setAttributes, isSelected } ) {
+				<EditorView
+					content={ attributes.content }
+					editorId={'block-library-html__editor'}
+					editorInstructionsId={'block-library-html__editor-instructions'}
```
I think we should use [`useInstanceId()`](https://github.com/WordPress/gutenberg/blob/4cfaab6302c862ce264247ebef31b45729f04343/packages/compose/src/hooks/use-instance-id/index.ts#L48) to generate unique ids in almost all cases, but especially here because there will very likely be multiple instances.

mirka commented on packages/block-editor/src/components/global-styles/editor-view.js:
```diff
@@ -0,0 +1,98 @@
+		} )();
+		// We only want to run this once, so we can ignore the dependency array.
+		// eslint-disable-next-line react-hooks/exhaustive-deps
+	}, [] );
```
I'm not sure this is tenable in a React context. I'm assuming from this code that the `content`, `onChange`, `onBlur`, and `mode` props are all basically initial config and cannot be dynamically updated. As a component consumer, that is unexpected to me. We could for example tweak the component API to make it clearer that they're initial config (like have consumers pass in the config through the output of a `useEditorView( config )` hook or something).

okmttdhr replied:
Thanks for your comment!

>I'm assuming from this code that the content, onChange, onBlur, and mode props are all basically initial config and cannot be dynamically updated

Yes, the approach was chosen to initialize `new CmEditorView` just once, allowing CodeMirror to handle subsequent updates, as I was to minimize complexity.

>As a component consumer, that is unexpected to me.

It makes sense to me! I have just updated the component API to use the term `initialConfig` to clarify that these properties are intended for initial setup only, but WDYT?

mirka commented on packages/block-editor/src/components/global-styles/editor-view.js:
```diff
@@ -0,0 +1,98 @@
+			<div
+				id={ editorInstructionsId }
+				className={ editorInstructionsId }
+				style={ { display: 'none' } }
```
We usually use the [`VisuallyHidden`](https://wordpress.github.io/gutenberg/?path=/docs/components-visuallyhidden--docs) component for things like this. I guess it isn't _technically_ necessarily in the `describedby` case, but I think we should prefer it for code reuse, clarity of intent, and the maybe also the [search angle](https://www.accessibility-developer-guide.com/examples/sensible-aria-usage/describedby/#text-not-searchable).

mirka commented on packages/block-editor/src/components/global-styles/editor-view.js:
```diff
@@ -0,0 +1,98 @@
+			>
+				{ editorInstructionsText }
+				{ __(
+					`Press Escape then Tab to move focus out of the editor.`
```
By the way, I feel like this keyboard nav hint is necessary for sighted users, too. Should we consider showing it?

Ideally, this would be done with the [`BaseControl` component](https://wordpress.github.io/gutenberg/?path=/docs/components-basecontrol--docs), via the `help` prop.

talldan replied:
I think it'd be good to consider simplifying the copy too, as lots of users won't know what 'focus' means.

"Press escape to finish editing" or similar (I'm not a copywriter though 😄 )

okmttdhr replied:
I've updated the component to utilize `<BaseControl>` for displaying the `help` text for Additional CSS. 🙂

(Note that, regarding the Custom HTML block, I kept it hidden, as `Esc+Tab` is a universal action to focus out of a block)

mirka commented on packages/block-editor/src/components/global-styles/advanced-panel.js:
```diff
@@ -37,40 +90,61 @@ export default function AdvancedPanel( {
+			return;
+		}
+		setCustomeCSS( ensureMinLines( inheritedValue?.css, EDITOR_ID ) );
+		// eslint-disable-next-line react-hooks/exhaustive-deps
```
It wasn't immediately clear to me why we're excluding `customCSS` from the deps array. (I try to avoid `exhaustive-deps` lint overrides as much as possible because it's such a nasty source of bugs! 😅)

mirka commented on packages/block-editor/src/components/global-styles/advanced-panel.js:
```diff
@@ -37,40 +90,61 @@ export default function AdvancedPanel( {
-				spellCheck={ false }
-			/>
+			<label
+				htmlFor={ EDITOR_ID }
```
Apparently a generic div is [not labelable](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/label#for). (Confirmed not labeled in the accessibility tree.)

mirka commented on packages/edit-site/src/components/global-styles/style.scss:
```diff
@@ -119,24 +119,33 @@
+			font-weight: 500;
+			line-height: 1.4;
+			text-transform: uppercase;
+		}
```
For styling, we can leverage `BaseControl`, or use [`BaseControl.VisualLabel`](https://wordpress.github.io/gutenberg/?path=/docs/components-basecontrol--docs#with-visual-label).

mirka commented on packages/block-editor/src/components/global-styles/editor-view.js:
```diff
@@ -0,0 +1,98 @@
+			>
+				{ editorInstructionsText }
+				{ __(
+					`Press Escape then Tab to move focus out of the editor.`
```
Can we actually make the editor field blur on <kbd>Escape</kbd>? Currently, it just maintains the focus outline and I can still type in the field. At the very least I think we need some visual feedback for the <kbd>Escape</kbd>.

okmttdhr replied:
I've taken a look at the current implementation, and it seems to work as intended on my end. Here's what it looks like:

https://github.com/WordPress/gutenberg/assets/5287479/35acd461-f07e-45ba-be86-1e1e8e04fcf1

I'm wondering if there might be specific conditions or environments where this behavior isn't consistent. Could you share more details about the issue you're encountering?

mirka commented on packages/block-library/src/html/edit.js:
For the editor in the Custom HTML block, we'll need to style it in a way that works across different color schemes. This is how it currently looks in a dark color scheme:

<img width="376" alt="Code completion popover has an illegible color combination" src="https://github.com/WordPress/gutenberg/assets/555336/883fe8ad-8385-4074-86bb-636a53cd54aa">

mirka reviewed (commented):
Great work, it would be really nice to have this feature!

I commented on things I noticed on the first pass, but before code specifics, we need to make sure there are no accessibility blockers. If we can make it work for everyone, great. If not, that might mean making it a toggleable option.

I think the accessibility-related notes I made are addressable, but I don't have enough familiarity with the text editing mechanics inside the editor itself. The code completion popover works as expected for me in VoiceOver, so that's good.

cc @andrewhayward @afercia @alexstine

okmttdhr committed:
Fix wrong package version

Co-authored-by: Lena Morita <lena@jaguchi.com>

okmttdhr committed:
Fix typo in variable name

okmttdhr committed:
Remove unnecessary condition in useEffect

okmttdhr committed:
Use @\mixins

okmttdhr committed:
Add VisuallyHidden component to hide editor instructions

okmttdhr committed:
Fix editor height isn't calculated properly

okmttdhr committed:
Use BaseControl for Additional CSS

okmttdhr committed:
Utilize useInstanceId

ellatrix commented on packages/block-library/src/html/edit.js:
```diff
@@ -73,11 +73,16 @@ export default function HTMLEdit( { attributes, setAttributes, isSelected } ) {
+						content: attributes.content,
+						onChange,
+						mode: 'html',
+					}}
```
I have a concern here: ideally we don't want to display editor UI inside the canvas, only content. Is there any way to do that with CodeMirror? So stuff like Autocomplete etc. should ideally be rendered outside the iframe so it can be styled as UI. I don't know if that would mean initialising outside the iframe and targeting a node in the canvas, or initialising inside the the iframe and setting a different area for UI.

ellatrix replied:
Another concern is that I'd expect however the code block is styled/syntax highlighted to be the same on the front-end. Is that possible?

ellatrix replied:
I guess I should read better: I just saw that this is for the HTML block and not the Code block, so these comments are not relevant. :)

afercia commented:
> I think the accessibility-related notes I made are addressable, but I don't have enough familiarity with the text editing mechanics inside the editor itself. The code completion popover works as expected for me in VoiceOver, so that's good.

From an accessibility perspective, I'd like to avoid to repeat all the concerns that were originally raised for the core implementation in https://core.trac.wordpress.org/ticket/12423. Basically, Codemirror drastically alters the DOM making it hardly readable with assistive technology. Also, the Codemirror UI isn't greatly accessible.

That's the reason why in https://core.trac.wordpress.org/ticket/12423 the final decision was that codemirror can be entirely disabled by users. This whole feature must honor the user preference `syntax_highlighting` in `wp-admin/profile.php`.

tyxla commented on .eslintrc.js:
```diff
@@ -76,6 +76,31 @@ const restrictedImports = [
 		message:
 			"edit-widgets is a WordPress top level package that shouldn't be imported into other packages",
 	},
+	{
+		name: 'codemirror',
+		message:
+			'Please use dynamic import (`import()`) instead since it is a large dependency.',
+	},
+	{
+		name: '@codemirror/commands',
+		message:
+			'Please use dynamic import (`import()`) instead since it is a large dependency.',
+	},
+	{
+		name: '@codemirror/lang-css',
+		message:
+			'Please use dynamic import (`import()`) instead since it is a large dependency.',
+	},
+	{
+		name: '@codemirror/lang-html',
+		message:
+			'Please use dynamic import (`import()`) instead since it is a large dependency.',
+	},
+	{
+		name: '@codemirror/view',
+		message:
+			'Please use dynamic import (`import()`) instead since it is a large dependency.',
+	},
```
This is a bit repetitive. Does it make sense to use the [`patterns` option](https://eslint.org/docs/latest/rules/no-restricted-imports) to combine them into a single restricted import?

tyxla commented on packages/block-editor/src/components/global-styles/advanced-panel.js:
```diff
@@ -1,18 +1,75 @@
+	 * (height of all the elements in the sidebar except the editor) + (height of the header) + (height of the footer)
+	 * Currently, it's desktop-optimized. 
+	 */
+	const MARGIN = 234 + 60 + 25;
```
If we're going to keep these static, can we move them up there in the module, right below the module imports? 

I'd also suggest splitting each into a separate constant, which will allow separate comments for each value, perhaps making it all a bit more readable.

tyxla commented on packages/block-editor/src/components/global-styles/advanced-panel.js:
```diff
@@ -1,18 +1,75 @@
+ */
+function ensureMinLines( content ) {
+	const MIN_LINES = 10;
+	const LINE_HEIGHT = 18.2; // Height of one line in the editor
```
Same as above in terms of where the constants live.

tyxla commented on packages/block-editor/src/components/global-styles/advanced-panel.js:
```diff
@@ -62,15 +118,29 @@ export default function AdvancedPanel( {
+					initialConfig={
+						{
+							callback: () => ensureMaxHeight(instanceId),
+							content: ensureMinLines( inheritedValue?.css ),
```
The optional chaining here hints that we could have an `undefined` passed to `ensureMinLines()`. But `ensureMinLines()` expects `content` to be defined as it does `content.split()`. Should we add some checks or coalesce to a default value?

tyxla commented on packages/edit-site/src/components/global-styles/style.scss:
```diff
@@ -119,24 +119,21 @@
+				@include input-style__focus();
+			}
+		}
+		.cm-editor {
```
I wonder if we can make things less specific here. Right now we have 4 levels:
1. `.edit-site-global-styles-screen-css`
2. `.components-v-stack`
3. `.cm-editor`
4. `.cm-line`

Do we need all of that nesting?

tyxla commented on packages/block-library/src/html/edit.js:
```diff
@@ -73,11 +73,16 @@ export default function HTMLEdit( { attributes, setAttributes, isSelected } ) {
-					placeholder={ __( 'Write HTML…' ) }
-					aria-label={ __( 'HTML' ) }
+				<EditorView
+					editorId={'block-library-html__editor'}
```
```suggestion
					editorId="block-library-html__editor"
```

tyxla commented on packages/block-library/src/html/edit.js:
```diff
@@ -73,11 +73,16 @@ export default function HTMLEdit( { attributes, setAttributes, isSelected } ) {
-					aria-label={ __( 'HTML' ) }
+				<EditorView
+					editorId={'block-library-html__editor'}
+					editorInstructionsText={__(
```
Spacing seems to be a bit off here. Linter didn't run for some reason.

tyxla commented on packages/block-library/src/html/edit.js:
```diff
@@ -73,11 +73,16 @@ export default function HTMLEdit( { attributes, setAttributes, isSelected } ) {
+				<EditorView
+					editorId={'block-library-html__editor'}
+					editorInstructionsText={__(
+						`This editor allows you to input your custom HTML.`
```
Seems to be no need for a template literal, it can be a quoted string.

```suggestion
						"This editor allows you to input your custom HTML."
```

tyxla commented on packages/block-library/src/html/edit.js:
```diff
@@ -73,11 +73,16 @@ export default function HTMLEdit( { attributes, setAttributes, isSelected } ) {
-					onChange={ ( content ) => setAttributes( { content } ) }
-					placeholder={ __( 'Write HTML…' ) }
-					aria-label={ __( 'HTML' ) }
+				<EditorView
```
This is a big change to the `html` block and may need a CHANGELOG mention.

jsnajdr commented:
I created #60362 to explore reorganizing the `@codemirror/*` imports so that the webpack chunk layout is simpler.

When we import the CodeMirror modules with individual dynamic imports, like:
```js
await import( 'codemirror' )
await import( '@codemirror/commands' )
await import( '@codemirror/view' )
await import( '@codemirror/lang-css' )
```
then webpack assumes that the imports are independent, i.e., any of the modules can be imported at any time, without importing the others. That's why it creates 7 independent chunks for CodeMirror modules:

<img width="451" alt="Screenshot 2024-04-02 at 13 38 47" src="https://github.com/WordPress/gutenberg/assets/664258/94138208-19c2-45ca-9d33-907a88935aba">

For example, `@codemirror/commands` and `@codemirror/autocomplete` are dependencies of `codemirror`, they have been loaded already by `import( 'codemirror' )`, but because they are also parts of other imports (i.e., `@codemirror/autocomplete` is a dependency of `@codemirror/lang-css` and others), they have their own chunks, to support independent loading and also to avoid duplication.

But when I move the imports to be synchronous in a separate `./code-mirror.js` file:
```js
import 'codemirror'
import '@codemirror/commands'
import '@codemirror/view'

async function importLanguage() {
  await import( '@codemirror/lang-css' )
}
```
and import the file dynamically with `import( './code-mirror' )`, then webpack is able to infer two things:
1. `@codemirror/commands` and `@codemirror/view` are always imported together with `codemirror`, they can be all bundled together.
2. `@codemirror/lang-css` is loaded only in a situation where `codemirror` is already loaded. So, if `lang-css` depends on `autocomplete`, it doesn't need to worry about it because it's been already loaded and webpack knows it. The chunk needs to contain only the minimal code to load `lang-css` on top of `codemirror`.

The result is that there are only three chunks, with minimal size and no code duplication:

<img width="830" alt="Screenshot 2024-04-02 at 12 43 12" src="https://github.com/WordPress/gutenberg/assets/664258/01f2a530-d4ea-437e-98e3-e38408f1df97">

Much better layout. I'm also specifying a `webpackChunkName` on each import, because otherwise the chunk file names will be numeric. This is how the production build of the current PR looks like:

<img width="857" alt="Screenshot 2024-04-02 at 12 53 45" src="https://github.com/WordPress/gutenberg/assets/664258/da1e49a9-7000-4fca-8a9d-256ac37e4d44">

okmttdhr commented:
> > I think the accessibility-related notes I made are addressable, but I don't have enough familiarity with the text editing mechanics inside the editor itself. The code completion popover works as expected for me in VoiceOver, so that's good.
> 
> From an accessibility perspective, I'd like to avoid to repeat all the concerns that were originally raised for the core implementation in [core.trac.wordpress.org/ticket/12423](https://core.trac.wordpress.org/ticket/12423). Basically, Codemirror drastically alters the DOM making it hardly readable with assistive technology. Also, the Codemirror UI isn't greatly accessible.
> 
> That's the reason why in [core.trac.wordpress.org/ticket/12423](https://core.trac.wordpress.org/ticket/12423) the final decision was that codemirror can be entirely disabled by users. This whole feature must honor the user preference `syntax_highlighting` in `wp-admin/profile.php`.

Thank you for the valuable resources, @afercia! 
I have reviewed the [previous](https://core.trac.wordpress.org/ticket/12423) [discussions](https://core.trac.wordpress.org/ticket/41872). It appears that many of the accessibility issues have been [addressed in this version](https://blog.jupyter.org/improving-the-accessibility-of-jupyter-6c695db518d3) in the CodeMirror 6 released in 2022. 
Nevertheless, to get aligned with it, it'd be nice to make syntax highlighting opt-out to respect the user preference, as it does for the Customizer.

okmttdhr commented:
> > This code should work even when we adopt importmap later. Once we turn all @wordpress/* packages (specifically block-editor in this case) into modules by using output.module=true, we can externalize codemirror by using [defaultRequestToExternalModule](https://github.com/WordPress/gutenberg/blob/7b38263aea571c0c804c85b19ced92c6dfe6289b/packages/dependency-extraction-webpack-plugin/lib/util.js#L78) in DependencyExtractionWebpackPlugin and register the script as a module by [WordPress/wordpress-develop#5818](https://github.com/WordPress/wordpress-develop/pull/5818) setting codemirror as a dependency module.
> 
> Thanks for starting this work, I don't think we need to wait for all the packages to be modules to be able to use import maps. Import maps is already supported in Core today and scripts can load modules so I think by making code mirror a module, we can start using it today.
> https://github.com/WordPress/gutenberg/pull/60155#issuecomment-2029267859

Thanks for your suggestion, @youknowriad! In reference to [Script Modules in 6.5](https://make.wordpress.org/core/2024/03/04/script-modules-in-6-5/), my understanding is that for utilization of a module (in this PR use case), it must be declared as a dependent of the consumer to be added into the `importmap`. In the following example, `@my-plugin/shared` module will be available to the `@my-plugin/entry` module:

```php
wp_register_script_module(
	'@my-plugin/shared',
	plugin_dir_url( __FILE__ ) . 'shared.js'
);

wp_enqueue_script_module(
	'@my-plugin/entry',
	plugin_dir_url( __FILE__ ) . 'entry.js',
	array( '@my-plugin/shared' )
);
```

Thus, `@wordpress/block-editor` should act as a module and an entry module in this scenario. Also, [currently](https://make.wordpress.org/core/2024/03/04/script-modules-in-6-5/#:~:text=Modules%20cannot%20depend%20on%20scripts%20and%20scripts%20cannot%20depend%20on%20modules), it seems that modules cannot depend on scripts, and scripts cannot depend on modules. 
Please feel free to offer corrections if my understanding is inaccurate.

okmttdhr committed:
Switch the editor's light/dark mode

youknowriad commented:
> my understanding is that for utilization of a module (in this PR use case), it must be declared as a dependent of the consumer to be added into the importmap.

Oh It's unfortunately that this slipped through the cracks, I think we should have a way to add a module to the import map without it being a dependency of another module that gets loaded manually in order to be used by scripts. @sirreal how do we solve this, it feels like a small thing to allow without necessarily making all scripts modules.

afercia commented:
> it'd be nice to make syntax highlighting opt-out to respect the user preference, as it does for the Customizer.

Just for compelteness, it does it also for the Theme File Editor and for the Plugin File Editor.

sirreal commented:
> @sirreal how do we solve this, it feels like a small thing to allow without necessarily making all scripts modules.

Scripts and modules are completely separate right now, there's no interoperability right now. I hope to have a good solution for WordPress 6.6. [Work is happening in this Trac Ticket](https://core.trac.wordpress.org/ticket/60647).

okmttdhr committed:
Merge branch 'trunk' into add/codemirror

okmttdhr commented:
📝 I plan to revisit this PR again in a while, as I currently don't have much time. But feel free to take this over if anyone is interested.

annezazu commented:
Tagging in a few folks who might be interested in taking this across the line for 6.7 cc @carolinan @Mamaduka @aristath @t-hamano. It would be awesome to see this finally added in.

sirreal commented:
> Scripts and modules are completely separate right now, there's no interoperability or interdependence right now. I hope to have a good solution for WordPress 6.6. [Work is happening in this Trac Ticket](https://core.trac.wordpress.org/ticket/60647).

I've created [this "Script Modules: Allow scripts to depend on modules" ticket](https://core.trac.wordpress.org/ticket/61500) for efforts to allow Scripts to depend on Script Modules.

sirreal commented:
I've created https://github.com/WordPress/gutenberg/pull/68191 on top of this PR to demonstrate how https://github.com/WordPress/wordpress-develop/pull/8024 allows classic scripts to depend on script modules.

If you'd like to help that work move forward, please leave feedback on https://github.com/WordPress/wordpress-develop/pull/8024.
